bpo-18857: urlencode: Add flag to prevent rendering NoneType values as literal string 'None'

jayaddison opened the pull request:
When a dictionary-like object with value(s) `None` is passed to [urllib.parse.urlencode](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode), query-string elements are generated containing the string literal 'None'.

For example (credit: [Joshua Johnston](https://bugs.python.org/issue18857#msg196314)):

```python
>>> from urllib import urlencode
>>> urlencode({'josh': None})
'josh=None'
```

This can cause recipients of query-string data generated in this manner to misinterpret and potentially store parsed 'None' string values as data, when in fact a null/absent value was intended.

This pull request introduces an optional (default-disabled) flag called `standalone_keys`; a developer may optionally set this flag to allow query-string parameters to be unambiguously serialized as 'standalone keys' -- indicating that no value is present.

https://bugs.python.org/issue18857

<!-- issue-number: [bpo-18857](https://bugs.python.org/issue18857) -->
https://bugs.python.org/issue18857
<!-- /issue-number -->

jayaddison committed:
bpo-18857: Add test coverage for desired behaviour

jayaddison committed:
bpo-18857: Split-out test_urlencode_sequences_trivial test case

jayaddison committed:
bpo-18857: Add test coverage on existing behaviour

jayaddison committed:
bpo-18857: Add feature-flag with no-op implementation

jayaddison committed:
bpo-18857: Extend existing test coverage: top-level NoneType value during sequence encoding

jayaddison committed:
bpo-18857: Conditionally render NoneType values as standalone keys

jayaddison committed:
Add blurb

jayaddison committed:
bpo-18857: Add standalone_keys parameter to urllib.parse.urlencode documentation

remilapeyre commented on Lib/test/test_urlparse.py:
```diff
@@ -890,13 +890,34 @@ def test_parse_qsl_max_num_fields(self):
+    def test_urlencode(self):
+        result = urllib.parse.urlencode({'a': 1, 'b': None})
+        # bpo-18857: urlencode of a None value uses the string 'None'
+        assert set(result.split('&')) == {'a=1', 'b=None'}
```
```suggestion
        self.assertEqual(result, 'a=1&b=None')
```

remilapeyre commented on Lib/test/test_urlparse.py:
```diff
@@ -890,13 +890,34 @@ def test_parse_qsl_max_num_fields(self):
+            query={'a': 1, 'b': None},
+            standalone_keys=True
+        )
+        assert set(result.split('&')) == {'a=1', 'b'}
```
```suggestion
        self.assertEqual(result, 'a=1&b')
```

remilapeyre commented on Lib/test/test_urlparse.py:
```diff
@@ -890,13 +890,34 @@ def test_parse_qsl_max_num_fields(self):
+        result = urllib.parse.urlencode({'a': [1, 2], 'b': (3, 4, 5), 'c': None}, True)
         # we cannot rely on ordering here
-        assert set(result.split('&')) == {'a=1', 'a=2', 'b=3', 'b=4', 'b=5'}
+        assert set(result.split('&')) == {'a=1', 'a=2', 'b=3', 'b=4', 'b=5', 'c=None'}
```
```suggestion
        self.assertEqual(result, 'a=1&a=2&b=3&b=4&b=5&c=None')
```

jayaddison replied:
Thanks!  Your suggestion here (and the related assertion suggestions on previous lines) look like good robustness and readability improvements.

I've applied them locally and tests pass, so I'll batch them up in GitHub and hopefully should be able to apply them as a single commit retaining your authorship.  Here goes..

remilapeyre commented on Lib/test/test_urlparse.py:
```diff
@@ -890,13 +890,34 @@ def test_parse_qsl_max_num_fields(self):
         # Sequence and object values.
-        result = urllib.parse.urlencode({'a': [1, 2], 'b': (3, 4, 5)}, True)
+        result = urllib.parse.urlencode({'a': [1, 2], 'b': (3, 4, 5), 'c': None}, True)
         # we cannot rely on ordering here
```
This comment is not correct anymore, now that dicts keep their insertion order we should be able to rely on ordering

jayaddison replied:
While this does seem true at the moment, the notes in the [Python 3.6 changelog](https://docs.python.org/3/whatsnew/3.6.html#new-dict-implementation) mention that the insertion order should currently be considered an implementation detail and is not (yet) safe to rely on.  It may be safer to keep the existing comment until there's a stronger guarantee around preserved insertion order.

jayaddison replied:
Ah, I [stand corrected](https://mail.python.org/pipermail/python-dev/2017-December/151283.html), sorry!  It looks like this behaviour was standardized in Python 3.7.  Given that, your suggestion sounds good - I'll take a moment to pause and read further before making applying this change.

remilapeyre replied:
Yes, in Python 3.6 the dict started to keep their insertion order but it was not official. It was made a part of the language in Python 3.7 and in Python 3.8 they were made reversable.

jayaddison replied:
Change applied; thank you.

remilapeyre commented on Lib/test/test_urlparse.py:
```diff
@@ -890,13 +890,34 @@ def test_parse_qsl_max_num_fields(self):
+            doseq=True,
+            standalone_keys=True
+        )
+        assert set(result.split('&')) == {'a=1', 'a', 'b'}
```
```suggestion
        self.assertEqual(result, 'a=1&a&b')
```

jayaddison committed:
bpo-18857: Simplify test assertions

- Remove string 'split' operation, enabling direct assertions on the test 'result' variable
- Use unittest assertEqual function in preference to builtin 'assert' keyword

Co-authored-by: Rémi Lapeyre <remi.lapeyre@henki.fr>

jayaddison committed:
bpo-18857: Remove redundant dictionary-ordering-related comment

remilapeyre commented on Lib/test/test_urlparse.py:
```diff
@@ -890,13 +890,33 @@ def test_parse_qsl_max_num_fields(self):
+        )
+        self.assertEqual(result, 'a=1&a&b')
 
+    def test_urlencode_sequences_trivial(self):
```
```suggestion
    def test_urlencode_sequences___str__(self):
```

remilapeyre commented on Misc/NEWS.d/next/Library/2020-05-05-20-55-31.bpo-18857.jFahmw.rst:
```diff
@@ -0,0 +1,5 @@
+Added ``standalone_keys`` flag (disabled-by-default) to
```
```suggestion
Added ``standalone_keys`` flag (disabled by default) to
```

remilapeyre commented on Misc/NEWS.d/next/Library/2020-05-05-20-55-31.bpo-18857.jFahmw.rst:
```diff
@@ -0,0 +1,5 @@
+Added ``standalone_keys`` flag (disabled-by-default) to
+``urllib.parse.urlencode``.
```
```suggestion
:meth:`urllib.parse.urlencode`.
```

remilapeyre reviewed (approved):
Thanks @jayaddison!

jayaddison committed:
bpo-18857: Improve unit test name

Co-authored-by: Rémi Lapeyre <remi.lapeyre@henki.fr>

jayaddison committed:
bpo-18857: Improve NEWS.d entry markup

Co-authored-by: Rémi Lapeyre <remi.lapeyre@henki.fr>
